encord/utilities/coco/exporter.py

```diff
@@ -0,0 +1,1031 @@
+try:
+    import pycocotools
+    import shapely
+except ImportError as e:
+    raise ImportError(
+        "The 'pycocotools' and 'shapely' packages are required for the COCO export. "
+        "Install them with: `pip install encord[coco]`"
+    ) from e
+
+import logging
+from collections import defaultdict
+from dataclasses import dataclass
+from itertools import chain
+from typing import Dict, List, Optional, Tuple, Union
+
+import numpy as np
+from pycocotools import mask as cocomask
+from pydantic import BaseModel
+from shapely.geometry import Polygon
+
+from encord.exceptions import EncordException
+from encord.objects.attributes import Attribute
+from encord.objects.common import PropertyType, Shape
+from encord.objects.ontology_object import Object
+from encord.objects.ontology_structure import OntologyStructure
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class DicomAnnotationData:
+    dicom_instance_uid: str
+    multiframe_frame_number: Optional[int]  # only present for multi-frame DICOMs
+    file_uri: str
+    width: int
+    height: int
+
+
+class CocoAnnotation(BaseModel):
+    area: float
+    bbox: Tuple[float, float, float, float]
+    category_id: int
+    id_: int
+    image_id: int
+    is_crowd: int
+    segmentation: Union[List, Dict]
+    keypoints: Optional[List[float]] = None
+    num_keypoints: Optional[int] = None
+    track_id: Optional[int] = None
+    encord_track_uuid: Optional[str] = None
+    rotation: Optional[float] = None
+    classifications: Optional[Dict] = None
+    manual_annotation: Optional[bool] = None
+
+    def to_dict(self) -> Dict:
+        return {
+            "area": self.area,
+            "bbox": list(self.bbox),
+            "category_id": self.category_id,
+            "image_id": self.image_id,
+            "iscrowd": self.is_crowd,
+            "segmentation": self.segmentation,
+            "keypoints": self.keypoints,
+            "num_keypoints": self.num_keypoints,
+            "id": self.id_,
+            "attributes": {
+                k: v
+                for k, v in {
+                    "track_id": self.track_id,
+                    "encord_track_uuid": self.encord_track_uuid,
+                    "rotation": self.rotation,
+                    "classifications": self.classifications,
+                    "manual_annotation": self.manual_annotation,
+                }.items()
+                if v is not None
+            },
+        }
+
+
+@dataclass
+class Size:
+    width: int
+    height: int
+
+
+def get_polygon_from_dict(
+    polygon_dict: Dict,
+    w: int,
+    h: int,
+) -> List[Tuple[float, float]]:
+    return [
+        (
+            polygon_dict[str(i)]["x"] * w,
+            polygon_dict[str(i)]["y"] * h,
+        )
+        for i in range(len(polygon_dict))
+    ]
+
+
+class CocoExporter:
+    """
+    This class is intentionally designed in a modular fashion to facilitate extensibility.
```

Comment:
Is this comment true? If so why?

Reply:
I remember that you brought it here from the original code, so probably you know the best 😄. From my pov as the docstring mentions, it "should" allow the user to override some methods to get a response better suited to their needs. Although "should" uses the "" as I don't have any record of an automated test for this where the original code is.